In [70]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [71]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [72]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [73]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [74]:
df_market_data.dtypes

price_change_percentage_24h     float64
price_change_percentage_7d      float64
price_change_percentage_14d     float64
price_change_percentage_30d     float64
price_change_percentage_60d     float64
price_change_percentage_200d    float64
price_change_percentage_1y      float64
dtype: object

In [75]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
df_market_data_scaled = StandardScaler().fit_transform(
    df_market_data[['price_change_percentage_24h', 'price_change_percentage_7d', 'price_change_percentage_14d',
                   'price_change_percentage_30d', 'price_change_percentage_60d', 'price_change_percentage_200d',
                   'price_change_percentage_1y']])

# Create a DataFrame with the scaled data

df_market_data_transformed = pd.DataFrame(df_market_data_scaled, columns=['price_change_percentage_24h', 'price_change_percentage_7d', 'price_change_percentage_14d',
                   'price_change_percentage_30d', 'price_change_percentage_60d', 'price_change_percentage_200d',
                   'price_change_percentage_1y'])

df_market_data_transformed.head()

price_change_percentage_24h  price_change_percentage_7d  \
0                     0.508529                    0.493193   
1                     0.185446                    0.934445   
2                     0.021774                   -0.706337   
3                    -0.040764                   -0.810928   
4                     1.193036                    2.000959   

   price_change_percentage_14d  price_change_percentage_30d  \
0                     0.772200                     0.235460   
1                     0.558692                    -0.054341   
2                    -0.021680                    -0.061030   
3                     0.249458                    -0.050388   
4                     1.760610                     0.545842   

   price_change_percentage_60d  price_change_percentage_200d  \
0                    -0.067495                     -0.355953   
1                    -0.273483                     -0.115759   
2                     0.008005                     -0.550247   
3                    -0.373164                     -0.458259   
4                    -0.291203                     -0.499848   

   price_change_percentage_1y  
0                   -0.251637  
1                   -0.199352  
2                   -0.282061  
3                   -0.295546  
4                   -0.270317

In [76]:
df_market_data.columns
#the coin_id column is set as the index of this dataframe

Index(['price_change_percentage_24h', 'price_change_percentage_7d',
       'price_change_percentage_14d', 'price_change_percentage_30d',
       'price_change_percentage_60d', 'price_change_percentage_200d',
       'price_change_percentage_1y'],
      dtype='object')

In [77]:
#get the values from the index of the dataframe, that will be the coin_id
crypto_names = df_market_data.index.values
crypto_names_df = pd.DataFrame(crypto_names)
crypto_names_df.head()

0
0       bitcoin
1      ethereum
2        tether
3        ripple
4  bitcoin-cash

In [78]:
# Copy the crypto names from the original data

df_market_data_transformed = pd.concat([df_market_data_transformed, crypto_names_df], axis=1)
df_market_data_transformed.set_index(0, inplace=True)


# Set the coinid column as index
df_market_data_transformed.rename_axis('coin_id', inplace=True)

# Display sample data
df_market_data_transformed.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [79]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [80]:
# Create an empty list to store the inertia values
inertia = []


# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
for i in k:
# 1. Create a KMeans model using the loop counter for the n_clusters
    k_model = KMeans(n_clusters=i, random_state=1)
    
# 2. Fit the model to the data using `df_market_data_scaled`
    k_model.fit(df_market_data_transformed)
# 3. Append the model.inertia_ to the inertia list    
    inertia.append(k_model.inertia_)





In [83]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)
df_elbow.head()

k     inertia
0  1  287.000000
1  2  195.820218
2  3  123.190482
3  4   79.022435
4  5   65.405923

In [84]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
df_elbow.hvplot.line(
    x="k", 
    y="inertia", 
    title="Elbow Curve", 
    xticks=k
)


:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [85]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=1)

In [86]:
# Fit the K-Means model using the scaled data
model.fit(df_market_data_transformed)

KMeans(n_clusters=4, random_state=1)

In [87]:
# Predict the clusters to group the cryptocurrencies using the scaled data
k_lower = model.predict(df_market_data_transformed)

# Print the resulting array of cluster values.
k_lower

array([3, 3, 1, 1, 3, 3, 3, 3, 3, 1, 1, 1, 1, 3, 1, 3, 1, 1, 3, 1, 1, 3,
       1, 1, 1, 1, 1, 1, 3, 1, 1, 1, 0, 3, 1, 1, 2, 1, 1, 1, 1],
      dtype=int32)

In [88]:
# Create a copy of the DataFrame
df_market_data_copy = df_market_data_transformed.copy()

In [89]:
# Add a new column to the DataFrame with the predicted clusters
df_market_data_copy['clusters_lower'] = k_lower

# Display sample data
df_market_data_copy.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  clusters_lower  
coin_id                                                   
bitcoin                        -0.251637               3  
ethereum                       -0.199352               3  
tether                         -0.282061               1  
ripple                         -0.295546               1  
bitcoin-cash                   -0.270317               3

In [90]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
df_market_data_copy.hvplot.scatter(
    x="price_change_percentage_24h",
    y="price_change_percentage_7d",
    by="clusters_lower",
    hover_cols="coin_id"
).opts(yformatter="%.0f")

:NdOverlay   [clusters_lower]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [91]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [92]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
crypto_pca = pca.fit_transform(df_market_data_transformed)

# View the first five rows of the DataFrame. 
crypto_pca[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [93]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** 0.3719856 + 0.34700813 + 0.17603793 = 0.89503166. 
The three principal components are able to capture 89.5% of the variance from our original data.

In [94]:
# Create a new DataFrame with the PCA data.
crypto_pca_df = pd.DataFrame(
    crypto_pca,
    columns=["PCA1", "PCA2", "PCA3"]
)
# Creating a DataFrame with the PCA data

crypto_pca_df.head()

PCA1      PCA2      PCA3
0 -0.600667  0.842760  0.461595
1 -0.458261  0.458466  0.952877
2 -0.433070 -0.168126 -0.641752
3 -0.471835 -0.222660 -0.479053
4 -1.157800  2.041209  1.859715

In [95]:
# Copy the crypto names from the original data
new_crypto_pca_df = pd.concat([crypto_pca_df, crypto_names_df], axis=1)

# Set the coinid column as index
new_crypto_pca_df.set_index(0, inplace=True)
new_crypto_pca_df.rename_axis('coin_id', inplace=True)

# Display sample data
new_crypto_pca_df.head()

PCA1      PCA2      PCA3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [96]:
# Create a list with the number of k-values from 1 to 11

k = list(range(1, 11))

In [97]:
# Create an empty list to store the inertia values
inertia_pca = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_pca`
# 3. Append the model.inertia_ to the inertia list

for i in k:
    k_model = KMeans(n_clusters=i, random_state=1)
    k_model.fit(new_crypto_pca_df)
    inertia_pca.append(k_model.inertia_)


In [98]:
# Create a dictionary with the data to plot the Elbow curve
pca_elbow_data = {"k": k, "inertia_pca": inertia_pca}

# Create a DataFrame with the data to plot the Elbow curve
pca_df_elbow = pd.DataFrame(pca_elbow_data)
pca_df_elbow.head()

k  inertia_pca
0  1   256.874086
1  2   165.901994
2  3    93.774626
3  4    49.665497
4  5    38.352251

In [101]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
pca_df_elbow.hvplot.line(
    x="k", 
    y="inertia_pca", 
    title="Elbow Curve", 
    xticks=k
)

:Curve   [k]   (inertia_pca)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:**4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** It is the same.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [102]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=0)

In [103]:
# Fit the K-Means model using the PCA data
model.fit(new_crypto_pca_df)

KMeans(n_clusters=4, random_state=0)

In [104]:
# Predict the clusters to group the cryptocurrencies using the PCA data
k_4 = model.predict(new_crypto_pca_df)

# Print the resulting array of cluster values.
k_4

array([0, 0, 3, 3, 0, 0, 0, 0, 0, 3, 3, 3, 3, 0, 3, 0, 3, 3, 0, 3, 3, 0,
       3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 1, 0, 3, 3, 2, 3, 3, 3, 3],
      dtype=int32)

In [105]:
# Create a copy of the DataFrame with the PCA data
new_crypto_pca_df_copy = new_crypto_pca_df.copy()

# Add a new column to the DataFrame with the predicted clusters
new_crypto_pca_df_copy['cluster_values'] = k_4

# Display sample data
new_crypto_pca_df_copy.head()

PCA1      PCA2      PCA3  cluster_values
coin_id                                                   
bitcoin      -0.600667  0.842760  0.461595               0
ethereum     -0.458261  0.458466  0.952877               0
tether       -0.433070 -0.168126 -0.641752               3
ripple       -0.471835 -0.222660 -0.479053               3
bitcoin-cash -1.157800  2.041209  1.859715               0

In [106]:
# Create a scatter plot using hvPlot by setting 
# `x="PCA1"` and `y="PCA2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
new_crypto_pca_df_copy.hvplot.scatter(
    x="PCA1",
    y="PCA2",
    by="cluster_values",
    hover_cols='coin_id'
)

:NdOverlay   [cluster_values]
   :Scatter   [PCA1]   (PCA2,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [121]:
# Composite plot to contrast the Elbow curves, create a composite dataframe
composite_df = pd.merge(df_elbow, pca_df_elbow, on='k')
composite_df.head()

k     inertia  inertia_pca
0  1  287.000000   256.874086
1  2  195.820218   165.901994
2  3  123.190482    93.774626
3  4   79.022435    49.665497
4  5   65.405923    38.352251

In [124]:
# Composite elbow curve plot

composite_df.hvplot.line(
    x="k", 
    y=["inertia", "inertia_pca"], 
    title="Elbow Curve", 
    xticks=k
)


:NdOverlay   [Variable]
   :Curve   [k]   (value)

In [130]:
# Composite plot to contrast the clusters
scatter1 = df_market_data_copy.hvplot.scatter(x="price_change_percentage_24h", y="price_change_percentage_7d")
scatter2 = new_crypto_pca_df_copy.hvplot.scatter(x="PCA1", y="PCA2")
composite_scatter = scatter1 + scatter2
composite_scatter
                                                 

:Layout
   .Scatter.I  :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d)
   .Scatter.II :Scatter   [PCA1]   (PCA2)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** Both resulted in the same k value, however with the PCA model the inertia is lower which would suggested that the data points are more tightly clustered together around the cluster center with a smaller standard deviation.